### PR TITLE
[RFC] Automatically invoke pytest_deselected for removed items

### PR DESCRIPTION
**Pull request by blueyed**
I remember @RonnyPfannschmidt suggesting this somewhere, but could not find it.

### CODE REVIEW

**Comment by blueyed** [comment]
I could also imagine emitting a warning instead only..

Came across this for when plugins rely on modified items being reported really for garbage collection (https://github.com/tarpas/pytest-testmon/pull/125).

File: src/_pytest/main.py [hunk 1]
```diff
@@ -491,14 +491,24 @@ def gethookproxy(self, fspath):
             proxy = self.config.hook
         return proxy
 
+    _deselected = []
```
Comment (RonnyPfannschmidt) [inline src/_pytest/main.py]: this **must** be a instance variable instead of a class one

Comment (blueyed) [reply src/_pytest/main.py]: Yeah.  As you can see from its placement this is just a hack / proof of concept anyway.

**Comment by RonnyPfannschmidt** [comment]
overall i like the idea, it sorts our a common issue/misstake

File: src/_pytest/main.py [hunk 2]
```diff
@@ -491,14 +493,26 @@ def gethookproxy(self, fspath):
...
+            items_before = set(items)
             hook.pytest_collection_modifyitems(
                 session=self, config=self.config, items=items
             )
+            ignore = set(self._deselected + items)
+            removed_items = items_before.difference(ignore)
+            if removed_items:
+                # Explicitly deselect them.
+                # Calls the own hook again, but that is OK.
+                self.config.hook.pytest_deselected(items=removed_items)
```
Comment (nicoddemus) [inline src/_pytest/main.py]: I guess with this change we can stop calling `pytest_deselected` explicitly in `cacheprovider`, `stepwise`, `mark`, etc?

Comment (blueyed) [reply src/_pytest/main.py]: Yes.

**Comment by blueyed** [comment]
@RonnyPfannschmidt 
Do you want it?  Otherwise feel free to close.

**Comment by blueyed** [comment]
Closing, feel free to re-open / merge, of course.

**Comment by RonnyPfannschmidt** [comment]
its impossible to reopen/merge now

**Comment by blueyed** [comment]
You can restore the branch, and re-open, no?

**Pull request closed without merging**